src/pages/VeSDL/OnChainVote.tsx

```diff
@@ -0,0 +1,254 @@
+  gauges: Partial<{
+    [lpTokenAddress: string]: Gauge
+  }>
+}
+export type GaugeName = {
+  address: string
+  gaugeName: string
+}
+
+export type VotesType = {
+  [gaugeAddress: string]: {
+    gaugeName: string
+    weight: BigNumber
+    voteDate: number
+  }
+}
+
+export default function OnChainVote({
+  veSdlBalance,
+  gauges,
+}: OnChainVoteProps) {
+  const theme = useTheme()
+  const { account } = useActiveWeb3React()
+  const { t } = useTranslation()
+  const [selectedGauge, setSelectedGauge] = useState<GaugeName | null>(null)
+  const [alertMessage, setAlertMessage] = useState<string>()
+  const [voteWeightToSubmit, setVoteWeightToSubmit] = useState<string>("")
+  const [voteUsed, setVoteUsed] = useState<number | undefined>()
+  const [votes, setVotes] = useState<VotesType>({})
+  const gaugeControllerContract = useGaugeControllerContract(true)
+  const [voteWeightInputError, setVoteWeightInputError] = useState<{
+    hasError: boolean
+    errorText: string
+  }>({ hasError: false, errorText: " " })
+
+  const gaugeNames = useMemo(
+    () =>
+      Object.keys(gauges)
+        .filter((lpAddress) => !gauges[lpAddress]?.isKilled)
+        .map((lpAddress) => {
+          const gaugeName = gauges[lpAddress]?.gaugeName
+          const gaugeAddress = gauges[lpAddress]?.address
+          if (gaugeName && gaugeAddress)
+            return {
+              address: gaugeAddress,
+              gaugeName: gaugeName,
+            } as GaugeName
+        }),
+    [gauges],
+  )
+
+  const getVoteUsed = useCallback(async () => {
+    if (account) {
+      const voteUsed = await gaugeControllerContract?.[
+        "vote_user_power(address)"
+      ](account)
+      setVoteUsed(voteUsed?.toNumber())
+    }
+  }, [account, gaugeControllerContract])
+
+  useEffect(() => {
+    if (account && gaugeControllerContract) {
+      gaugeControllerContract.on(
+        "VoteForGauge",
+        (voteDate, fromAddress, gaugeAddress, voteWeight) => {
+          const gaugeName = gaugeNames.find(
+            (gaugeName) =>
+              gaugeName?.address.toLowerCase() ===
+              (gaugeAddress as string).toLowerCase(),
+          )?.gaugeName
+
+          if (
+            gaugeName &&
+            (fromAddress as string).toLowerCase() === account.toLowerCase()
+          ) {
+            setVotes((currentVoteLists) => ({
+              ...currentVoteLists,
+              [gaugeAddress]: {
+                gaugeName,
+                voteDate: (voteDate as BigNumber).toNumber(),
+                weight: voteWeight as BigNumber,
+              },
+            }))
+            void getVoteUsed()
+          }
+        },
+      )
+    }
+    return () => {
+      gaugeControllerContract?.removeAllListeners()
+    }
+  }, [gaugeControllerContract, gaugeNames, votes, getVoteUsed, account])
+
+  const handleVote = async () => {
+    const voteWeightPercent = parseFloat(voteWeightToSubmit)
+    if (!Number.isNaN(voteWeightPercent)) {
```

Comment:
Why would this not be a number?

Reply:
This checks `Nan` value of `voteWeightPercent`

Reply:
a percent is always a number so it should always be between 0-1 which is why checking if it's a number is weird to me from reading standpoint

Reply:
`voteWeightToSubmit` is text we input, `parseFloat('abcedf')` returns `Nan`, so it is worth to check this value whether it is `Nan` or not.
Also, FYI, the percent number is in 0~100% in here.

Reply:
it will be better to do `number` attribute on that input or use regex to only allow numbers